File: tests/bench/queens/Queens.java
// Same program as queens.jai.
public class Queens {
    // BENCH_LEVEL picks how much work this benchmark does: easy a sixteenth of it,
    // medium a quarter, hard (the default, and anything unrecognised) all of it.
    static long scale() {
        String l = System.getenv("BENCH_LEVEL");
        if ("easy".equals(l)) return 16;
        if ("medium".equals(l)) return 4;
        return 1;
    }

    static final long SCALE = scale();
    // Scale the repetitions, not the board: work explodes with n.
    static final long REPS = Math.max(1L, 3L / SCALE);

    static boolean safe(long[] cols, long row, long col) {
        for (long r = 0; r < row; r++) {
            long c = cols[(int) r];
            if (c == col) return false;
            if (c - r == col - row) return false;
            if (c + r == col + row) return false;
        }
        return true;
    }

    static long place(long[] cols, long row, long n) {
        if (row == n) return 1;
        long found = 0;
        for (long col = 0; col < n; col++) {
            if (safe(cols, row, col)) {
                cols[(int) row] = col;
                found += place(cols, row + 1, n);
            }
        }
        return found;
    }

    public static void main(String[] args) {
        final int n = 11;
        long[] cols = new long[n];
        long total = 0;
        for (long rep = 0; rep < REPS; rep++) total = place(cols, 0, n);
        System.out.println(n);
        System.out.println(total);
    }
}

File: tests/bench/queens/queens.cpp
// Same program as queens.jai. See tests/bench/README.md for why the C++ and
// Java rows are here and what they are not.
#include <cstdint>
#include <cstdio>
#include <vector>

#include <cstdlib>
#include <cstring>

// BENCH_LEVEL picks how much work this benchmark does: easy a sixteenth of it,
// medium a quarter, hard (the default, and anything unrecognised) all of it.
static long bench_scale() {
    const char *l = std::getenv("BENCH_LEVEL");
    if (l && std::strcmp(l, "easy") == 0) return 16;
    if (l && std::strcmp(l, "medium") == 0) return 4;
    return 1;
}
static const long SCALE = bench_scale();
// A repetition count around a fixed board: never below one round.
static const long REPS = (3 / SCALE) < 1 ? 1 : (3 / SCALE);

static bool safe(const std::vector<int64_t> &cols, int64_t row, int64_t col) {
    for (int64_t r = 0; r < row; r++) {
        int64_t c = cols[(size_t)r];
        if (c == col) return false;
        if (c - r == col - row) return false;
        if (c + r == col + row) return false;
    }
    return true;
}

static int64_t place(std::vector<int64_t> &cols, int64_t row, int64_t n) {
    if (row == n) return 1;
    int64_t found = 0;
    for (int64_t col = 0; col < n; col++) {
        if (safe(cols, row, col)) {
            cols[(size_t)row] = col;
            found += place(cols, row + 1, n);
        }
    }
    return found;
}

int main() {
    const int64_t n = 11;
    std::vector<int64_t> cols((size_t)n, 0);
    int64_t total = 0;
    for (long rep = 0; rep < REPS; rep++) total = place(cols, 0, n);
    std::printf("%lld\n", (long long)n);
    std::printf("%lld\n", (long long)total);
    return 0;
}
